linker/shared/src/main/scala/org/scalajs/linker/backend/emitter/FunctionEmitter.scala

```diff
@@ -357,12 +357,12 @@ private[emitter] class FunctionEmitter(jsGen: JSGen) {
 
     var syntheticVarCounter: Int = 0
 
-    def newSyntheticVar()(implicit pos: Position): Ident = {
-      /* TODO Integrate this with proper name management.
-       * This is filed as #2971.
-       */
+    @tailrec
+    private def newSyntheticVar()(implicit pos: Position): Ident = {
       syntheticVarCounter += 1
-      Ident("jsx$" + syntheticVarCounter, None)
+      val varName = "jsx$" + syntheticVarCounter
+      if (localVarNames.contains(varName)) newSyntheticVar()
```

Comment:
What if the user-defined local var is encountered *after* we already decided to allocate `jsx$1` to a synthetic one?

Reply:
Do you mean something like this?

```scala
while (idx >= 0 && name.charAt(idx) != '$') {
  idx -= 1
  var `jsx$1` = "" + name
  buf += `jsx$1`
}
```

If so, this case passes in this PR.

Reply:
The below is fine too.
```scala
while (idx >= 0 && name.charAt(idx) != '$') {
  idx -= 1
  var `jsx$1` = "" + name // user-defined `jsx$1` after synthetic one
  var idx2 = name.length - 2
  while (idx2 >= 0 && name.charAt(idx2) != '$') {
    idx2 -= 1
    buf += `jsx$1`
  }
}
```